plugins/providers/hyperv/cap/configure_disks.rb

```diff
@@ -0,0 +1,200 @@
+require "log4r"
+require "fileutils"
+require "vagrant/util/numeric"
+require "vagrant/util/experimental"
+
+module VagrantPlugins
+  module HyperV
+    module Cap
+      module ConfigureDisks
+        LOGGER = Log4r::Logger.new("vagrant::plugins::hyperv::configure_disks")
+
+        # @param [Vagrant::Machine] machine
+        # @param [VagrantPlugins::Kernel_V2::VagrantConfigDisk] defined_disks
+        # @return [Hash] configured_disks - A hash of all the current configured disks
+        def self.configure_disks(machine, defined_disks)
+          return {} if defined_disks.empty?
+
+          return {} if !Vagrant::Util::Experimental.feature_enabled?("disks")
+
+          machine.ui.info(I18n.t("vagrant.cap.configure_disks.start"))
+
+          current_disks = machine.provider.driver.list_hdds
+
+          configured_disks = {disk: [], floppy: [], dvd: []}
+
+          defined_disks.each do |disk|
+            if disk.type == :disk
+              disk_data = handle_configure_disk(machine, disk, current_disks)
+              configured_disks[:disk] << disk_data unless disk_data.empty?
+            elsif disk.type == :floppy
+              # TODO: Write me
+              machine.ui.info(I18n.t("vagrant.cap.configure_disks.floppy_not_supported", name: disk.name))
+            elsif disk.type == :dvd
+              # TODO: Write me
+              machine.ui.info(I18n.t("vagrant.cap.configure_disks.dvd_not_supported", name: disk.name))
+            end
+          end
+
+          configured_disks
+        end
+
+        protected
+
+        # @param [Vagrant::Machine] machine - the current machine
+        # @param [Config::Disk] disk - the current disk to configure
+        # @param [Array] all_disks - A list of all currently defined disks in VirtualBox
+        # @return [Hash] current_disk - Returns the current disk. Returns nil if it doesn't exist
+        def self.get_current_disk(machine, disk, all_disks)
+          current_disk = nil
+          if disk.primary
+            # Ensure we grab the proper primary disk
+            # We can't rely on the order of `all_disks`, as they will not
+            # always come in port order, but primary should always be Location 0 Number 0.
+
+            current_disk = all_disks.select { |d| d["ControllerLocation"] == 0 && d["ControllerNumber"] == 0 }.first
```

Comment:
You can replace the `select...first` expression with a `detect`

Reply:
I think that might make things more complicated, as `all_disks` is an array of hashes. I'd have to map a `detect` on each hash in the array, right? `select` will apply the enumerator block to each hash and return the result. The `.first` is kinda gross, but there should only be 1 disk that matches (and select always returns the results in an array).

Reply:
Hmm, it's not working like I expected it would in pry, so I'll bring up a vm and try it.

Reply:
Alright, switched it over to detect. Dunno why it was working differently in a pry session but whatever! haha